geode-redis/src/main/java/org/apache/geode/redis/internal/executor/string/SetExecutor.java

```diff
@@ -35,105 +37,182 @@
+      existsOption = SetOptions.Exists.XX;
+    }
     return new SetOptions(existsOption, expiration, keepTTL);
   }
 
-  private long parseExpirationTime(int index, List<byte[]> commandElems)
-      throws IllegalArgumentException {
-    String expirationString;
+  private String getNextParameter(String currentParameter,
+      List<String> optionalParametersStrings) {
+    int index = optionalParametersStrings.indexOf(currentParameter);
+    if (optionalParametersStrings.size() <= index + 1) {
+      throw new IllegalArgumentException(ERROR_SYNTAX);
+    }
+    return optionalParametersStrings.get(index + 1);
+  }
+
+  private void throwExceptionIfUnknownParameter(List<String> optionalParameters) {
+    List<String> validOptionalParamaters = Arrays.asList("EX", "PX", "NX", "XX", "KEEPTL");
+
+    List<String> parametersInQuestion =
+        optionalParameters
+            .stream()
+            .filter(parameter -> (!validOptionalParamaters.contains(parameter)))
+            .collect(Collectors.toList());
+
+    parametersInQuestion.forEach(parameter -> throwErrorIfNotANumberInExpectedPosition(
+        parameter,
+        optionalParameters));
+  }
+
+  private void throwErrorIfNotANumberInExpectedPosition(
+      String parameter,
+      List<String> optionalParameters) {
+
+    if (previousOptionIsValidAndExpectsANumber(parameter, optionalParameters)) {
+      convertToLongOrThrowException(parameter);
+    } else {
+      throw new IllegalArgumentException(ERROR_SYNTAX);
+    }
+  }
 
+  private boolean isANumber(String parameter) {
     try {
-      expirationString = Coder.bytesToString(commandElems.get(index));
-    } catch (IndexOutOfBoundsException e) {
+      Long.parseLong(parameter);
+      return true;
+    } catch (NumberFormatException e) {
+      return false;
+    }
+  }
+
+  private boolean previousOptionIsValidAndExpectsANumber(String paramter,
+      List<String> optionalParameters) {
+
+    List<String> validParamaters = Arrays.asList("EX", "PX");
+    if (optionalParameters.size() < 2) {
+      return false;
+    }
+
+    int indexOfParameter = optionalParameters.indexOf(paramter);
+    String previousParameter = optionalParameters.get(indexOfParameter - 1);
+
+    return validParamaters.contains(previousParameter);
+  }
+
+
+  private void throwExceptionIfIncompatableParamaterOptions(List<String> passedParametersStrings) {
+
+    if (passedParametersStrings.contains("PX")
+        && passedParametersStrings.contains("EX")) {
       throw new IllegalArgumentException(ERROR_SYNTAX);
     }
 
+    if (passedParametersStrings.contains("XX")
+        && passedParametersStrings.contains("NX")) {
+      throw new IllegalArgumentException(ERROR_SYNTAX);
+    }
+  }
+
+  private long parseExpirationTime(String expirationParameter, String timeUntilExpiration) {
+
+    long millisecondsUntilExpirationLong =
+        Long.parseLong(timeUntilExpiration);
+
+    if (expirationParameter.equals("EX")) {
```

Comment:
I think it would be better to do this in the caller that does this for EX. Then you don't need to pass expirationParameter to this method. Its job is just parsing.

Reply:
ok-  done (when commit arrives)